File: 2019/02/part1.py
#!/usr/bin/python3
# https://adventofcode.com/2019/day/2

def main():
    data = open("input.txt").read()
    mem = [ int(part) for part in data.split(",") if part.isdigit() ]

    mem[1] = 12
    mem[2] = 2

    ptr = 0

    while mem[ptr] != 99:
        opcode = mem[ptr]
        op0 = mem[ptr + 1]
        op1 = mem[ptr + 2]
        dst = mem[ptr + 3]

        if opcode == 1: # add
            mem[dst] = mem[op0] + mem[op1]
        elif opcode == 2: # multiply
            mem[dst] = mem[op0] * mem[op1]

        ptr += 4

    print(mem[0])

main()

File: 2019/02/part1.cpp
#include <iostream>

#include "../../utils/utils.h"


using namespace std;


int main()
{
    auto lines = read_lines("input.txt");
    auto parts = split(lines[0], ',');
    auto ram = to_int(parts);

    size_t ptr = 0;

    ram[1] = 12;
    ram[2] = 2;

    while(true) {
        int opcode = ram[ptr];
        if (opcode == 99) {
            break;
        }

        switch(opcode) {
            case 1:
                ram[ram[ptr + 3]] = ram[ram[ptr + 1]] + ram[ram[ptr + 2]];
                break;
            case 2:
                ram[ram[ptr + 3]] = ram[ram[ptr + 1]] * ram[ram[ptr + 2]];
                break;
        }
        ptr += 4;
    }

    cout << ram[0] << endl;
    return 0;
}